### ISSUE TITLE
LM API changes for stabilization and future flexibility

### ISSUE THREAD

**Issue by reuben** [mozilla/DeepSpeech#2672]
**Problem:** Currently we have implementation assumptions coded into the LM API. For example, the `DS_EnableDecoderWithLM` call assumes there are two separate files, one for the LM binary and one for the trie. It also assumes there are two float hyperparameters alpha and beta. If we keep it as is, then any change in the underlying external scorer technology requires breaking the API.

**Complication:** We don't currently have a concrete idea of how the underlying scorer implementation would change, so we don't know what the future API would need.

**Proposal:** Any problem in computer science can be solved with another layer of indirection. Separate the decoder loading and the hyperparameter setting into separate API calls so the latter can be made deprecated/no-op in case we move to a new implementation where it isn't needed. Figure out a way to package our current KenLM solution into a single file instead of separate LM.binary and trie files. We can then iterate on the underlying file format and handle upgrades without breaking the API, for example by supporting two implementations at the same time until a future 2.x release.

The new API would look something like this:

```cpp
int DS_EnableExternalScorer(ModelState* ctx, const char* scorerModelPath);
int DS_SetScorerAlphaBeta(ModelState* ctx, float alpha, float beta);
```

Ideally we can push default alpha and beta values into the LM package so that most users don't have to call `DS_SetScorerAlphaBeta` (makes upgrading easier). For an even more flexible approach, where we can deprecate individual options and add new options without breaking the API, we can also do something like:

```cpp
enum DS_ScorerOptions
{
    DS_SCORER_ALPHA,
    DS_SCORER_BETA,
};

int DS_SetScorerOptionInt(ModelState* ctx, int optionCode, int optionValue);
int DS_SetScorerOptionFloat(ModelState* ctx, int optionCode, float optionValue);
int DS_SetScorerOptionStr(ModelState* ctx, int optionCode, char* optionValue);
```

**Implementation details:** KenLM takes file paths everywhere for loading the LM binary file, so encapsulating it in an archive is hard to do without patching KenLM. What we can do instead is append the trie and any extra data/configuration to the end of the LM binary file, making a single file that still loads with the KenLM API and we can then load the trie from the correct offset. We can then embed default alpha/beta values in the trie file header, which we already control.

This is hacky, but the goal here is just to move to a single file, so that we can iterate on the underlying binary file format without breaking the API. A future solution could use a cleaner packaging solution, but the point is that it's invisible to API users.

- [x] Update KenLM vendored code
- [x] Stop including vocabulary data in LM.binary.
- [x] Refactor Scorer so that trie generation can be performed by `generate_trie` directly.
- [x] Make `generate_trie` take the vocabulary directly as input instead of extracting it from the LM.
- [x] Adapt `generate_trie` to generate new package format as a single file.
- [x] Adapt Scorer to load the new format.
- [x] Include default alpha and beta hyperparameters in trie header.
- [x] Refactor DeepSpeech API to load a single path and make hyperparameter setting optional.
- [x] Adapt all clients and bindings.
- [x] Improve error handling around scorer loading.

**Comment by reuben** [comment]
@lissyx @kdavis-mozilla @tilmankamp I'd appreciate any feedback you have on the API proposal. In particular, do you think the second option is unnecessarily abstract?

**Comment by tilmankamp** [comment]
I prefer the second option. For a minute I thought about an (API-wise even simpler) parameter-string alternative, but this could easily end up in an implementation nightmare.

**Comment by kdavis-mozilla** [comment]
**Proposal**: For the suggested API
```cpp
int DS_SetScorerOptionInt(ModelState* ctx, int optionCode, int optionValue);
int DS_SetScorerOptionFloat(ModelState* ctx, int optionCode, float optionValue);
int DS_SetScorerOptionStr(ModelState* ctx, int optionCode, char* optionValue);
```
people can use it incorrectly (Think of the discourse posts!) doing things like
```cpp
DS_SetScorerOptionInt(ctx, DS_SCORER_ALPHA, 3);
```
Using and enum in this way, as far as I understand, will only _not_ break the ABI if we only add new values, i.e.
```cpp
enum DS_ScorerOptions
{
    DS_SCORER_ALPHA,
    DS_SCORER_BETA,
    DS_SCORER_FOO,
};
```
If we were to ever  remove values
```cpp
enum DS_ScorerOptions
{
    DS_SCORER_FOO,
};
```
we would break the ABI. 

However, we'd only remove values if we dropped support for older LM techniques. Are we OK with requiring an ABI break when we drop support for an old LM technique?

Also, a nit, the string should be immutable
```cpp
int DS_SetScorerOptionStr(ModelState* ctx, int optionCode, const char* optionValue);
```
**Implementation details**: Would something less fancy work?

For example zipping the LM, trie, and a config file (containing alpha + beta) and unzipping, or putting some code in front of KenLM to make it looks like normal files on disk.

If we do something like you suggest we'll have to create tooling, and documentation, to allow others to create such "packed" language models.

**Comment by lissyx** [comment]
> people can use it incorrectly (Think of the discourse posts!) doing things like

That's what would worry me the most.

> However, we'd only remove values if we dropped support for older LM techniques. Are we OK with requiring an ABI break when we drop support for an old LM technique?

IMHO I'd agree it's an acceptable compromise here.



> For example zipping the LM, trie, and a config file (containing alpha + beta) and unzipping, or putting some code in front of KenLM to make it looks like normal files on disk.

I think that zipping is going to be much painful, adding an extra step and thus latency, even if it's hidden by the API.



> This is hacky, but the goal here is just to move to a single file

Totally agree with that, even if it's hacky. Though, given your proposal and especially `DS_EnableExternalScorer` I think we should ensure this is actually working and that there is no snarky issue that would arise. One of the thing I'm wondering is `mmap()`. As much as I recall, we don't use it for the `trie` file, but it's being used by KenLM.

Basically, I think your proposal might be a good improvement, but I'm (generally speaking) not a big big fan of that kind of interface (I hate it for getopt). I also don't have a better proposal for the moment.

**Comment by reuben** [comment]
> **Proposal**: For the suggested API
> 
> ```c++
> int DS_SetScorerOptionInt(ModelState* ctx, int optionCode, int optionValue);
> ```
> 
> people can use it incorrectly (Think of the discourse posts!) doing things like
> 
> ```c++
> DS_SetScorerOptionInt(ctx, DS_SCORER_ALPHA, 3);
> ```

Yes, this is why I included an explicit option:

```c++
int DS_SetScorerAlphaBeta(ModelState* ctx, float alpha, float beta);
```

We would then deprecate the function rather than an individual entry in an enum, and make it a no-op if we ship a new LM implementation that doesn't have alpha and beta hyperparameters.

---

> If we were to ever remove values
> [...]
> we would break the ABI.
> 
> However, we'd only remove values if we dropped support for older LM techniques. Are we OK with requiring an ABI break when we drop support for an old LM technique?

I can't think of any way to remove support for a feature without breaking the API. We should be planning with this in mind, whatever we ship in 1.x will stay until 2.x. We can introduce a new option in 1.x and deprecate old options, but not remove them.

---

> **Implementation details**: Would something less fancy work?
> 
> For example zipping the LM, trie, and a config file (containing alpha + beta) and unzipping, or putting some code in front of KenLM to make it looks like normal files on disk.

Putting some code in front of KenLM to trick it would be much more complicated, we'd have to somehow emulate POSIX file APIs, it sounds like a recipe for disaster to me.

> If we do something like you suggest we'll have to create tooling, and documentation, to allow others to create such "packed" language models.

It's not so complicated:

```bash
# generate LM.binary
# generate LM.trie
cat LM.binary LM.trie > LM.package
```

---

> One of the thing I'm wondering is mmap(). As much as I recall, we don't use it for the trie file, but it's being used by KenLM.

I've tested that the approach above (concatenating the files) works for KenLM, I now need to figure out how to get the total LM file size from KenLM to skip to the correct offset for the trie. And we also use `mmap` for the trie.

> Basically, I think your proposal might be a good improvement, but I'm (generally speaking) not a big big fan of that kind of interface (I hate it for getopt). I also don't have a better proposal for the moment.

OK, so for now the no abstraction option is the favorite one. I'll continue working on getting the single file working for now, and after that's working we can revisit the hyperparameter option if needed.

**Comment by reuben** [comment]
> It's not so complicated:
> 
> ```shell
> # generate LM.binary
> # generate LM.trie
> cat LM.binary LM.trie > LM.package
> ```

It is actually more complicated. Will need to make a tool (e.g. adapt `generate_trie`) to generate the packed format.

**Comment by reuben** [comment]
Got it running and producing correct results after loading both the LM and the trie from the same file.

**Comment by kdavis-mozilla** [comment]
Dumb question. As this is more-or-less working, mod clean up and testing, do you want to put it in 0.7.0 which will give time for others to "kick the tires" on it before it's in 1.0.0?

**Comment by reuben** [comment]
Sure, I think it's likely this will get done before the other items in 0.7.

**Comment by reuben** [comment]
All changes merged. See #2698
